src/test/java/co/rsk/bitcoinj/core/TransactionWitnessTest.java

```diff
@@ -361,10 +364,290 @@ public void getSigInsertionIndex_withWitnessFilledWithSignatures_shouldReturnThe
+        assertSignaturesAreInOrder(witnessWithTwoSignatures, Lists.newArrayList(fedKey1TxSignature, fedKey1TxSignature));
+    }
+
+    @Test
+    public void updateWitnessWithSignature_withWitnessFilledWithSignatures_shouldThrowAnError() {
+        // signing order: [fedKey1, .., maxFedKey, fedKey1]
+        // arrange
+        int i = 0;
+        while(i < redeemScript.getNumberOfSignaturesRequiredToSpend()) {
+            BtcECKey key = FEDERATION_KEYS.get(i);
+            signInput(btcTx, key, FIRST_INPUT_INDEX, btcTxSigHashForWitness);
+            i++;
+        }
+
+        // getSuffixCount doesn't consider OP_NOTIF, so the calculation for
+        // the number of signatures required in updateWitnessWithSignature is
+        // wrong.
+        BtcECKey key = FEDERATION_KEYS.get(i++);
+        signInput(btcTx, key, FIRST_INPUT_INDEX, btcTxSigHashForWitness);
+
+        byte[] txSig = getTransactionSignatureEncodedToBtc(fedKey1, btcTxSigHashForWitness);
+        TransactionWitness signedTransactionWitness = btcTx.getWitness(FIRST_INPUT_INDEX);
+        key = FEDERATION_KEYS.get(i);
+        int sigInsertionIndex = signedTransactionWitness.getSigInsertionIndex(btcTxSigHashForWitness, key);
+
+        // act & assert
+        assertThrows(IllegalArgumentException.class, () -> signedTransactionWitness.updateWitnessWithSignature(
+            p2shP2wshOutputScript, txSig, sigInsertionIndex));
+    }
+
+    @Test
+    public void updateWitnessWithSignature_withTheLowestSignatureInWitness_shouldInsertTheNewOneAsSecond() {
+        // signing order: [fedKey1, fedKey2]
+        byte[] fedKey1SignatureEncoded = getTransactionSignatureEncodedToBtc(fedKey1, btcTxSigHashForWitness);
+        TransactionWitness witnessWithoutSignatures = btcTx.getWitness(FIRST_INPUT_INDEX);
+        int fed1SigInsertionIndex = witnessWithoutSignatures.getSigInsertionIndex(btcTxSigHashForWitness, fedKey1);
+
+        // act & assert
+        TransactionWitness witnessWithFedKey1Signature = witnessWithoutSignatures.updateWitnessWithSignature(
+            p2shP2wshOutputScript, fedKey1SignatureEncoded, fed1SigInsertionIndex);
+
+        assertSignaturesAreInOrder(witnessWithFedKey1Signature, Lists.newArrayList(fedKey1SignatureEncoded));
+
+        byte[] fedKey2TxSignature = getTransactionSignatureEncodedToBtc(fedKey1, btcTxSigHashForWitness);
+        int fed2SigInsertionIndex = witnessWithFedKey1Signature.getSigInsertionIndex(btcTxSigHashForWitness, fedKey2);
+        TransactionWitness witnessWithTheSameSignatureTwice = witnessWithFedKey1Signature.updateWitnessWithSignature(
+            p2shP2wshOutputScript, fedKey2TxSignature, fed2SigInsertionIndex);
+
+        assertSignaturesAreInOrder(witnessWithTheSameSignatureTwice, Lists.newArrayList(fedKey1SignatureEncoded, fedKey1SignatureEncoded));
+    }
+
+    @Test
+    public void updateWitnessWithSignature_withALowerSignatureInWitness_shouldInsertTheNewOneAsFirst() {
+        // signing order: [fedKey2, fedKey1]
```

Comment:
I'd make this comment a bit more clear, show the order in which we are signing and also the order on how the signatures should be arranged in the witness script